docs/design/upgrade-resource-state.md

```diff
@@ -0,0 +1,581 @@
+
+### New VersionedResourceType Interface
+
+The framework could wrap the existing `ResourceType` type with a versioning interface, then requiring that provider developers fully define every version of a `ResourceType` and `Resource` when implementing a managed resource. The `Version` field of the `Schema` type would be removed to prevent conflicting code.
+
+For example:
+
+```go
+// New type
+type VersionedResourceType interface {
+    GetResourceTypeVersions(context.Context) map[int64]ResourceType
+}
+
+// Existing type
+type Provider interface {
+    // Required method update from map[string]ResourceType
+    GetResources(context.Context) (map[string]VersionedResourceType, diag.Diagnostics)
+}
+
+// Existing type
+type Resource interface {
+    // New required method
+    // If there is no previous version, a blank implementation would be required.
+    UpgradeState(context.Context, /*...*/) /*...*/
+}
+```
+
+This would ensure framework resources:
+
+- Contain all schema and logic necessary over time to perform any resource state upgrades, which simplifies the framework implementation.
+- Do not require additional provider developer discovery for resource versioning features.
+- Have access to the provider client during upgrade state operations.
+
+However, there are quite a few drawbacks to this type of implementation:
+
+- Additional and generally unnecessary complexity for most provider developers.
+- Lots of code build up over time unless provider developers find ways to reduce it.
+- Existing framework provider implementations would require updates to `GetResources` and each `Resource`.
+- Confusing blank `(Resource).UpgradeState()` implementation for single version resources. It could be made as an optional method, but then provider developers need to discover it and know when to use it, removing a main benefit of this type of approach.
+
+### New ResourceTypeWithUpgrades Interface
+
+The framework could allow provider developers to optionally extend their `ResourceType` with a new `UpgradeState` method.
+
+For example in the framework:
+
+```go
+// New type
+type ResourceTypeWithUpgradeState interface {
+    UpgradeState(context.Context, /*...*/) /*...*/
+}
+```
+
+Which would result in the following optional provider implementation:
+
+```go
+func (rt ExampleResourceType) UpgradeState(ctx context.Context, /*...*/) /*...*/ {
+    /* ... */
+}
+```
+
+However, there is an immediate drawback that there would be no access to the provider client during upgrade state operations. This is considered a non-starter due to the goals of this functionality.
```

Comment:
Although the headings are different, this text is identical to the following section - _New ResourceWithUpgradeState Interface_.

Reply:
Hey sorry, I'm not sure what you mean here. 😅  `ResourceType` and `Resource` are very different implementation details. Providers get the opportunity to inject their clients into `Resource`, but not `ResourceType`.

Reply:
My bad 😄